design/src/rfcs/rfc0040_behavior_major_versions.md

```diff
@@ -0,0 +1,89 @@
+
+For new SDKs, a GA cutline presents difficult choices around timeline and features that can’t be added later without altering behavior.
+
+Both of these use cases are addressed by Behavior Major Versions.
+
+<!-- The "Terminology" section is optional but is really useful for defining the technical terms you're using in the RFC -->
+Terminology
+-----------
+
+- **Some Term**: A definition for that term
+
+<!-- Explain how users will use this new feature and, if necessary, how this compares to the current user experience -->
+The user experience if this RFC is implemented
+----------------------------------------------
+
+In the current version of the SDK, users can construct clients without indicating any sort of behavior major version.
+Once this RFC is implemented, there will be two ways to set a behavior major version:
+
+1. In code via `aws_config::from_env_with_version(BehaviorMajorVersion::latest())` and `<service>::Config::builder().behavior_major_version(...)`. This will also work for `config_override`.
+2. By enabling `behavior-version-latest` in either `aws-config` (which brings back `from_env`) OR a specific generated SDK crate
+
+```toml
+# Cargo.toml
+[dependencies]
+aws-config = { version = "1", features = ["behavior-version-latest"] }
+# OR
+aws-sdk-s3 = { version = "1", features = ["behavior-version-latest"] }
+```
+
+If no `BehaviorMajorVersion` is set, the client will panic during construction.
+
+`BehaviorMajorVersion` is an opaque struct with initializers like `::latest()`, `::v2023_11_09()`. Downstream code can check the version by calling methods like `::supports_v1()`
+
+When new BMV are added, the previous version constructor will be marked as `deprecated`. This serves as a mechanism to alert customers that a new BMV exists to allow them to upgrade.
+
+How to actually implement this RFC
+----------------------------------
+
+In order to implement this feature, we need to create a `BehaviorMajorVersion` struct, add config options to `SdkConfig` and `aws-config`, and wire it throughout the stack.
+```rust
+/// Behavior major-version of the client
+///
+/// Over time, new best-practice behaviors are introduced. However, these behaviors might not be backwards
+/// compatible. For example, a change which introduces new default timeouts or a new retry-mode for
+/// all operations might be the ideal behavior but could break existing applications.
+#[derive(Debug, Clone)]
+pub struct BehaviorMajorVersion {
+    // currently there is only 1 MV so we don't actually need anything in here.
+}
+```
+
+To help customers migrate, we are including `from_env` hooks that set `behavior-version-latest` that are _deprecated_. This allows customers to see that they are missing the required cargo feature and add it to remove the deprecation warning.
+
+Internally, `BehaviorMajorVersion` will become an additional field on `<client>::Config`. It is _not_ ever stored in the `ConfigBag` or in `RuntimePlugins`.
```

Comment:
Would be nice to add how the BMV will be incorporated into RuntimePlugins via constructors and separate config bag values.

Reply:
done